packages/lesswrong/components/comments/CommentPool.tsx

```diff
@@ -0,0 +1,298 @@
+
+  const context: CommentPoolContextType = useMemo(() => ({
+    showMoreChildrenOf, showParentOf, showAncestorChain, invalidateComment, addComment
+  }), [showMoreChildrenOf, showParentOf, showAncestorChain, invalidateComment, addComment]);
+  
+  const wrappedLoadMoreTopLevel = useCallback(async () => {
+    await loadAll();
+    stateRef.current = revealTopLevel(stateRef.current, 10);
+    forceRerender();
+  }, [forceRerender, loadAll]);
+  
+  const loadedComments: SingleCommentState[] = stateRef.current.commentsSortOrder.map(commentId => stateRef.current.commentsById[commentId]);
+  const visibleComments = filter(loadedComments, c=>c.visibility!=="hidden");
+  const tree = unflattenComments(visibleComments.map(c=>c.comment));
+
+  treeOptions = {
+    ...treeOptions,
+    singleLineCommentsShowDescendentCount: true,
+  };
+
+  return <CommentPoolContext.Provider value={context}>
+    {tree.map(comment =>
+      <CommentNodeOrPlaceholder
+        treeOptions={treeOptions}
+        startThreadTruncated={startThreadTruncated}
+        expandAllThreads={expandAllThreads}
+        treeNode={comment}
+        key={comment._id}
+        parentCommentId={parentCommentId}
+        parentAnswerId={parentAnswerId}
+        shortform={(treeOptions.post as PostsBase)?.shortform}
+        isChild={defaultNestingLevel > 1}
+      />
+    )}
+    {/*topLevelCommentCount && loadMoreTopLevel && topLevelCommentCount>tree.length*/
+      (
+        (!haveLoadedAll && topLevelCommentCount && topLevelCommentCount > countVisibleTopLevelComments(stateRef.current))
+          || hasHiddenTopLevelComments(stateRef.current)
+      ) && <LoadMore
+        loadMore={wrappedLoadMoreTopLevel}
+        count={tree.length}
+        totalCount={topLevelCommentCount}
+      />
+    }
+  </CommentPoolContext.Provider>
+}
+
+function initialStateFromComments(initialComments: CommentsList[]): CommentPoolState {
+  return {
+    commentsSortOrder: initialComments.map(c=>c._id),
+    commentsById: keyBy(
+      initialComments.map(comment => ({comment, visibility: "visible"})),
+      c => c.comment._id
+    )
+  };
+}
+
+function addLoadedComments(state: CommentPoolState, loadedComments: CommentsList[]): CommentPoolState {
+  const newCommentStates = {...state.commentsById};
+  const addedCommentIds: string[] = [];
+  for (let comment of loadedComments) {
+    if (!newCommentStates[comment._id]) {
+      addedCommentIds.push(comment._id);
+      newCommentStates[comment._id] = {
+        comment, visibility: "hidden"
+      };
+    }
+  }
+  return {
+    ...state,
+    commentsSortOrder: [...state.commentsSortOrder, ...addedCommentIds],
+    commentsById: newCommentStates,
+  };
+}
+
+function countVisibleTopLevelComments(state: CommentPoolState): number {
+  let count = 0
+  for (let commentState of Object.values(state.commentsById)) {
+    if (!commentState.comment.parentCommentId && commentState.visibility!=="hidden") {
+      count++
+    }
+  }
+  return count
```

Comment:
nit: Can this be a `Object.values(state.commentsById).filter(...).length`?  also lets you extract the logic to a named function, i.e. `isVisibleTopLevelComment`, or something.